### PR TITLE
Lazyload supports auto managing all svc ports

### PR DESCRIPTION
**Pull request by cywang1905**
## Meaning

The previous Lazyload required a manual update of `general.wormholePort` in SlimeBoot, and redeployment when the service port need managed changed.

The Lazyload module now supports **automatic management** of all service ports. When a user creates or updates a service, the Lazyload module automatically determines if the range of service ports has changed, and includes the new ports in the management.

## Details

1. In order to correctly initialize the resources associated with the global-sidecar component, `general.wormholePort` **must not be empty at startup**.

2. Compatible with previous versions, ports manually specified in `general.wormholePort` still be managed by Lazyload.
3. For security reasons, the port will **only be increased automatically**. If you need to reduce the range of ports, you need to manually modify `general.wormholePort` in SlimeBoot.

4. Lazyload follows Istio's logic for port protocols (port names starting with "HTTP, HTTP2, GRPC, GRPCWeb") and only senses port changes for **HTTP protocols**.

5. The name of the SlimeBoot should be **the same as** the name of the main module, otherwise the SlimeBoot associated with the Lazyload cannot be found.

6. Services in the Istio root namespace are not managed (Specified via SlimeBoot `global.istioNamespace`, default is `istio-system`)

## Usage
1.Initially configure Lazyload to manage only port 80
```yaml
kind: SlimeBoot
metadata:
  name: lazyload
  namespace: mesh-operator
spec:
  # ...
  module:
    - name: lazyload
      kind: lazyload
      enable: true
      general:
        wormholePort: # at least one
          - "80"
      global:
        misc:
          globalSidecarMode: cluster # inform the mode of global-sidecar
          metricSourceType: accesslog # indicate the metric source
 # ...
```
3. Create a new http service on port 9080, such as the bookinfo demo service

4. Lazyload includes port 9080 in management
```yaml
      # ...
      general:
        wormholePort: # at least one
          - "80"
          - "9080"
      # ...
```

---
## 意义

之前的懒加载，在需要纳管的服务端口发生变化时，需要手动更新SlimeBoot中的`general.wormholePort`，并重新部署。
现在懒加载模块支持**自动纳管**所有服务端口。当用户创建或更新服务时，懒加载模块会自动判断服务端口范围是否变化，并将新增端口纳入管理。

## 细节

1. 为了正确初始化兜底组件global-sidecar的相关资源，启动时`general.wormholePort`**不能为空**

2. 兼容之前版本，`general.wormholePort`中手动指定的端口都会被懒加载纳管
3. 为了安全，纳管端口**只会自动增加**，如果需要缩减端口范围，需要手动修改SlimeBoot中`general.wormholePort`

4. 懒加载遵循Istio对端口协议的判断逻辑（端口名以“HTTP, HTTP2, GRPC, GRPCWeb”开头），只感知**HTTP协议**的端口变化

5. **SlimeBoot的名字要与主模块名字相同**，否则无法找到懒加载相关的SlimeBoot

7. Istio根命名空间的服务不纳管（通过SlimeBoot `global.istioNamespace`指定，默认是`istio-system`）

## 用法

1. 初始时配置懒加载只管理80端口
```yaml
kind: SlimeBoot
metadata:
  name: lazyload
  namespace: mesh-operator
spec:
  # ...
  module:
    - name: lazyload
      kind: lazyload
      enable: true
      general:
        wormholePort: # at least one
          - "80"
      global:
        misc:
          globalSidecarMode: cluster # inform the mode of global-sidecar
          metricSourceType: accesslog # indicate the metric source
 # ...
```

2. 新建9080端口的http服务，比如bookinfo demo的服务

4. 懒加载感知服务新增，将9080端口纳入管理
```yaml
      # ...
      general:
        wormholePort: # at least one
          - "80"
          - "9080"
      # ...
```

### CODE REVIEW

File: staging/src/slime.io/slime/modules/lazyload/controllers/service_cache.go [hunk 1]
```diff
@@ -115,8 +117,93 @@ func newSvcCache(clientSet *kubernetes.Clientset) (*NsSvcCache, *LabelSvcCache,
...
+	wormPortMap := make(map[string]bool)
+
+	for _, p := range wormholePort {
+		wormPortMap[p] = true
+	}
+
+	for port, proto := range portProtocolCache.Data {
+		p := strconv.Itoa(int(port))
+		if proto[ProtocolHTTP] > 0 && !wormPortMap[p] {
+			add = append(add, p)
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/lazyload/controllers/service_cache.go:191]: only inc?

File: staging/src/slime.io/slime/modules/lazyload/controllers/update_resources.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	// update slimeBoot
+	obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&slimeBoot)
+	if err != nil {
+		log.Errorf("convert slimeboot %s/%s to unstructured error: %+v", slimeBootNs, slimeBootName, err)
+		return
+	}
+	utd.SetUnstructuredContent(obj)
+	utd, err = dynCli.Resource(slimeBootGvr).Namespace(slimeBootNs).Update(utd, metav1.UpdateOptions{})
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/lazyload/controllers/update_resources.go:123]: no resourcever semantic? may override sth.

Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/lazyload/controllers/update_resources.go:123]: may cause frequent restart of lazyload pod

File: staging/src/slime.io/slime/modules/lazyload/controllers/update_resources.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		log.Errorf("convert slimeboot %s/%s to structured error: %v", slimeBootNs, slimeBootName, err)
+		return
+	}
+
+	for _, module := range slimeBoot.Spec.Modules {
+		if module.Kind != "lazyload" {
+			continue
+		}
+		// update wormholePort
+		module.General.WormholePort = wormholePort
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/lazyload/controllers/update_resources.go:113]: we can do once more equals-check (remember to order it firstly)

File: staging/src/slime.io/slime/modules/lazyload/controllers/update_resources.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+var (
+	slimeBootGvr = schema.GroupVersionResource{
+		Group:    "config.netease.com",
+		Version:  "v1alpha1",
+		Resource: "slimeboots",
+	}
+)
+
+type SlimeBoot struct {
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/lazyload/controllers/update_resources.go:22]: a little dirty to re-declare these types

File: staging/src/slime.io/slime/modules/lazyload/controllers/service_cache.go [hunk 5]
```diff
@@ -83,13 +61,37 @@ func newSvcCache(clientSet *kubernetes.Clientset) (*NsSvcCache, *LabelSvcCache,
...
+			if ns != r.env.Config.Global.IstioNamespace {
+				portProtocolCache.Lock()
+				for _, port := range service.Spec.Ports {
+					p := port.Port
+					if portProtocolCache.Data[p] == nil {
+						continue
+					}
+					switch getProtocol(port) {
+					case ProtocolHTTP:
+						portProtocolCache.Data[p][ProtocolHTTP]--
```
[the remark this answers is not in the document]
Comment (cywang1905) [reply staging/src/slime.io/slime/modules/lazyload/controllers/service_cache.go:75]: want to implement port auto delete at the beginning, but now port can only be added, so these code will be removed

**Pull request closed without merging**